### PR TITLE
WIP: Add relay support

### PR DESCRIPTION
**Pull request by gauravtiwari**
Hello @justin808 

Added relay rendering support to this gem. This is work in progress and I wanted to get your thoughts on this. Could you check and let me know if this is something that could be part of this gem? 

Here is the demo blog: https://relay-rails-blog.herokuapp.com/, I deployed using your gem (but modified). Previously I was using `react-rails`, but I feel for full fledged frontend development workflow react-on-rails is much more structured and flexible. 

G

[Review on Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/285)

### CODE REVIEW

**Comment by justin808** [comment]
@gauravtiwari OMG! This sounds **GREAT**!!! Let's jump on a Screen Hero or Skype and discuss this! Send me your email and I'll add you to our slack channel.

Here's some things we need:

0. Create some docs in the README.md and in docs/additional_reading
1. Prototype a simple example in spec/dummy
2. Create capybara tests that verify the feature in spec/dumm
3. Create JS and Ruby unit tests.

**Comment by justin808** [comment]
Here's the demo app: https://github.com/gauravtiwari/relay-rails-blog

**Comment by gauravtiwari** [comment]
@justin808 Awesome! Sure thing, I will send you an email on your github public email. Yes, that's the demo blog application.

**Comment by justin808** [comment]
@gauravtiwari Looking forward to getting this feature going.

**Comment by justin808** [comment]
@gauravtiwari How's this going? We need CI to work.

**Comment by justin808** [comment]
@gauravtiwari Any updates?

**Comment by gauravtiwari** [comment]
Hey @justin808, The feature is almost complete (here is the blog that's running latest code: https://relay-rails-blog.herokuapp.com/). 

There was a bug introduced in Relay library and it just got fixed in the new version so, I am going to take a look at the integration later today. A couple of things to note is that the relay itself is not complete yet, a lot of things like `server side rendering and realtime subscriptions` are not supported and the API might change very soon. I am not sure if this is ok?

**Comment by justin808** [comment]
@gauravtiwari Get as far as you can and we'll check this out!

**Comment by justin808** [comment]
@gauravtiwari I'm going to soon merge my changes for v4 into master. It would be great to find out if your changes are compatible or if you have any requests.

https://github.com/shakacode/react_on_rails/pull/311/files

**Comment by justin808** [comment]
@gauravtiwari Let me know when I should take a look at this.

File: lib/react_on_rails/version_checker.rb [hunk 1]
```diff
@@ -57,7 +57,7 @@ def initialize(package_json)
       end
 
       def raw
-        JSON.parse(package_json_contents)["dependencies"]["react-on-rails"]
+        JSON.parse(package_json_contents)["dependencies"]["react-relay-on-rails"]
```
Comment (jbhatab) [inline lib/react_on_rails/version_checker.rb:60]: should we maybe still keep the name to be react-on-rails or make a separate library for react-on-rails-relay?

**Comment by jbhatab** [comment]
@gauravtiwari awesome work. Love this. so you are saying they don't support server rendering? What would that mean exactly for this?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-:-KF81kiJk6BftSpqO6-3:-1506816157)*

**Comment by jbhatab** [comment]
*[package.json, line 67 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF821zgIUR-y1t4d1lL:-KF821zgIUR-y1t4d1lM:-193004653) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/package.json#L67)):*
So are we making 2 npm packages or merging into one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[node_package/src/clientStartup.js, line 70 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF82JQpTArH6pCskafY:-KF82JQpTArH6pCskafZ:388677826) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/node_package/src/clientStartup.js#L70)):*
how does this work for nested routes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[node_package/src/ReactOnRails.js, line 151 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF8329OM6XIpVxAx8V-:-KF8329Pa8JlXj27sci3:1589845426) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/node_package/src/ReactOnRails.js#L151)):*
Little confused why there are two functions that call each other?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[node_package/src/serverRenderRelayComponent.js, line 14 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF83RKHpU41sW7PsdJk:-KF83RKHpU41sW7PsdJl:1434266843) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/node_package/src/serverRenderRelayComponent.js#L14)):*
Curious how things like renderLoading work with server rendering.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/Gemfile.lock, line 127 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF849Ss7hjh1TF1SoJK:-KF849StAsMTwXi6dhov:-1043727960) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/Gemfile.lock#L127)):*
just pointing out merge conflicts

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/Gemfile.lock, line 166 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF84D3UGWsZb73YgPsz:-KF84D3VAN517OgQUTvN:-1043727960) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/Gemfile.lock#L166)):*
just pointing out merge conflicts

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[node_package/src/serverRenderRelayComponent.js, line 20 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF85A087iP3PxUdbBb1:-KF85A09aNe4cFSUg_G0:-2038142564) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/node_package/src/serverRenderRelayComponent.js#L20)):*
So we can either bake this into the 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/app/views/pages/relay.html.erb, line 3 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF85dU9GrPsUmYbuhGI:-KF85dU9GrPsUmYbuhGJ:727115858) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/app/views/pages/relay.html.erb#L3)):*
Can we just infer route from the url since we are already passing that in? Not 100% sure how route is used in relay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/client/package.json, line 45 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF85nwI3wARlkdd7DsX:-KF85nwI3wARlkdd7DsY:1971783612) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/client/package.json#L45)):*
@justin808 @gauravtiwari how should we go about treating this in the generator? we are already looking into doing two things related to this. 1. Breaking out a component generator, but would we want relay/redux to stack on top of each other? 2. Combining the client/server webpack config into one file. You said that relay may not support server rendering so im not sure how this would play out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/client/app/startup/ServerReduxSharedStoreApp.jsx, line 5 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF86W7OdUJKeUMlDrE2:-KF86W7OdUJKeUMlDrE3:900986538) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/client/app/startup/ServerReduxSharedStoreApp.jsx#L5)):*
I think we may want to still keep this react-on-rails. I don't care too much though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/config/application.rb, line 12 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF86hSmOZ0Gp7T_UdDx:-KF86hSnkpe_Hfdc-p3u:1295565417) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/config/application.rb#L12)):*
Hmm since we have to add code like this to the application rb, we would need to sync up the relay option on the installer generator to the component generator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/config/application.rb, line 14 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF86qDs-58hOGLCmatN:-KF86qDs-58hOGLCmatO:2100958777) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/config/application.rb#L14)):*
Do we need this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/config/application.rb, line 16 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF86sg1x6F8qm49oRzy:-KF86sg1x6F8qm49oRzz:-1170593202) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/config/application.rb#L16)):*
Can we make the first line a wildcard?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[spec/dummy/config/application.rb, line 5 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF871nNEGBT7x6bpYhC:-KF871nNEGBT7x6bpYhD:1846679990) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/spec/dummy/config/application.rb#L5)):*
What does this do exactly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
*[node_package/src/serverRenderRelayComponent.js, line 20 \[r5\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/285#-KF85A087iP3PxUdbBb1:-KF85Fdujl5Z3pJgZtG2:454377460) ([raw file](https://github.com/shakacode/react_on_rails/blob/e3dde5ded8332822ebbbb3c49def941381e77f22/node_package/src/serverRenderRelayComponent.js#L20)):*
woops mispost. 

So is this going to be baked into the js api for react-on-rails? I feel like this may be better suited to just manually write it out. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by jbhatab** [comment]
Reviewed 4 of 8 files at r1, 17 of 24 files at r3, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/285)*

**Comment by gauravtiwari** [comment]
Hey @jbhatab Thanks for looking into this. The problem is, the Relay API is quite unstable at the moment, they just released 0.8.0 version without much documentation and a lot has changed since the previous versions. Server rendering support is a WIP. 

Due to these reasons I went quiet on this pull request. I think it would make sense to wait until API is stable (we could provide some docs for integration). Perhaps, a separate repo with couple of generators to add Relay Support for meantime?

There are also other GraphQL clients that does similar job. For example: https://github.com/apollostack/apollo-client and https://github.com/kadirahq/lokka and https://github.com/gyzerok/adrenaline (Redux specific)

Let me know what you think?

**Comment by jbhatab** [comment]
@gauravtiwari I agree, I just looked into some options and it's all over the place. I think just waiting to see how the community picks up things is the best. We may be able to pull off all of them somehow or not. Not sure yet. But I think waiting to see what formulates is best. It reminds me of all the flux implementations before redux.

**Comment by justin808** [comment]
@gauravtiwari Any updates on this open PR?

**Comment by ramusus** [comment]
Hi, guys! I spent few days trying to get this PR working and finally realized, we don't need it at all.
I just implemented support for relay on top of react-on-rails with react-router just using original 6.0.3 version. PR for tutorial will come in few days after I will learn relay mutations.

**Comment by jbhatab** [comment]
@ramusus Awesome!!! sounds perfect.

**Comment by justin808** [comment]
@ramusus I'd love to hear more about this! Please do share. Please [send me your email](mailto: justin@shakacode.com) if you'd like to join our slack discussion room for React on Rails.

**Comment by zdne** [comment]
@ramusus 

> PR for tutorial will come in few days after I learn relay mutations. 

any progress on this?

**Comment by ramusus** [comment]
@zdne sorry, a lot of work last month. I hope I make it ready till the end of July

**Comment by justin808** [comment]
@ramusus How's this one going?

**Comment by justin808** [comment]
@ramusus Any good examples of using this?

**Comment by gauravtiwari** [comment]
Hey @justin808, Saw your comment on medium. I am currently not working on this feature. As discussed earlier, it seems we shouldn't make relay part of this gem. It would add a lot of complexity and also supporting relay or any other GraphQL client would be a app specific requirement. 

For your reference, I did one simple implementation for apollo and react here: 

https://github.com/gauravtiwari/graphql-apollo-blog/blob/master/app/assets/javascripts/renderer.es6.js

and for relay here (with your library): 

https://github.com/gauravtiwari/react_on_rails/blob/master/node_package/src/clientStartup.js#L60

As you would notice, both requires a different setup, unlike react where we just render a component with some data. I suggest we drop this PR. What do you think?

**Comment by justin808** [comment]
Closing for now, per comments from @gauravtiwari

**Comment by ramusus** [comment]
@justin808 @zdne Just submitted https://github.com/shakacode/react-webpack-rails-tutorial/pull/328

**Comment by gl2748** [comment]
@gauravtiwari @justin808 
Ok so, general question here, but why would anyone implement GraphQL and then consume it with a React app all within a rails app? Keen to hash this out, as I see advantages but not sure if it's an anti-pattern type situation, curious to hear some expert opinions ....

**Comment by justin808** [comment]
@gl2748 My team has been focused on using Redux, rather than GraphQL.

**Comment by gl2748** [comment]
Hey @justin808 thanks for the response,
So as I understand it you're saying Redux is preferable for managing React component state than Relay from within a Rails app?

**Comment by justin808** [comment]
We never tried Relay.

**Pull request closed without merging**